File: ddw/refine.py
"""Endpoint-guided coordinate ascent over a fixed window schedule.
Backward messages encode the entire remaining schedule, not a short horizon.
"""

import argparse
import hashlib
import json
import math
import time
from pathlib import Path

import cupy as cp
import numpy as np

from .core import Window
from .gpu import Engine


def windows_at(initial, schedule, completed):
    left = schedule[completed - 1] if completed else initial[0]
    right = schedule[completed - 2] if completed >= 2 else initial[1 - completed]
    return left, right


def endpoint_value(prob, left, right, endpoint):
    i, j = left.index(endpoint[0]), right.index(endpoint[1])
    return float(prob[i, j]) if i is not None and j is not None else 0.0


def backward_messages(engine, initial, schedule, endpoint, cache="device"):
    count = len(schedule)
    left, right = windows_at(initial, schedule, count)
    terminal = cp.zeros((1 << len(left.bits), 1 << len(right.bits)), dtype=cp.float64)
    i, j = left.index(endpoint[0]), right.index(endpoint[1])
    if i is None or j is None:
        raise ValueError("endpoint is outside terminal windows")
    terminal[i, j] = 1
    messages = [None] * (count + 1)
    messages[count] = (cp.asnumpy(terminal) if cache == "host" else terminal, 0.0)
    for completed in range(count - 1, -1, -1):
        left, right = windows_at(initial, schedule, completed)
        prob = engine.adjoint(
            cp.asarray(messages[completed + 1][0]), left, right, schedule[completed]
        )
        peak = float(prob.max())
        if peak <= 0:
            raise ValueError("endpoint has no retained paths")
        prob *= 1 / peak
        messages[completed] = (
            cp.asnumpy(prob) if cache == "host" else prob,
            messages[completed + 1][1] + math.log2(peak),
        )
    return messages


def candidate_score(
    engine, prob, left, right, schedule, position, candidate, messages, endpoint
):
    """Three transitions restore common state windows; suffix contraction is exact."""
    stop = min(position + 3, len(schedule))
    value = prob
    scale = 0.0
    for index in range(position, stop):
        target = candidate if index == position else schedule[index]
        value, (peak, _, _) = engine.step_and_summary(value, left, right, target)
        if peak <= 0:
            return -math.inf
        value *= 1 / peak
        scale += math.log2(peak)
        left, right = target, left
    if stop == len(schedule):
        mass = endpoint_value(value, left, right, endpoint)
    else:
        backward, backward_scale = messages[stop]
        mass = float(cp.sum(value * cp.asarray(backward)))
        scale += backward_scale
    return scale + math.log2(mass) if mass > 0 else -math.inf


def optimize(
    engine,
    initial,
    schedule,
    endpoint,
    width,
    candidates,
    passes,
    report=None,
    cache="device",
):
    schedule = list(schedule)
    history = []
    for sweep in range(passes):
        messages = backward_messages(engine, initial, schedule, endpoint, cache)
        before = endpoint_value(
            messages[0][0], *initial, (initial[0].base, initial[1].base)
        )
        before = messages[0][1] + math.log2(before)
        prob = cp.ones((1, 1), dtype=cp.float64)
        left, right = initial
        prefix_scale = 0.0
        accepted = 0
        for position, original in enumerate(schedule):
            proposed = engine.candidates(
                prob, right, engine.basis(left), width, count=candidates
            )
            choices = [original] + [window for window in proposed if window != original]
            stop = min(position + 3, len(schedule))
            saved_message = messages[stop]
            if stop < len(schedule) and isinstance(saved_message[0], np.ndarray):
                messages[stop] = (cp.asarray(saved_message[0]), saved_message[1])
            best, best_score = original, -math.inf
            original_score = None
            for candidate in choices:
                score = candidate_score(
                    engine,
                    prob,
                    left,
                    right,
                    schedule,
                    position,
                    candidate,
                    messages,
                    endpoint,
                )
                if original_score is None:
                    original_score = score
                if score > best_score + 1e-12:
                    best, best_score = candidate, score
            messages[stop] = saved_message
            if best != original:
                accepted += 1
                change = dict(
                    sweep=sweep + 1,
                    round=position + 1,
                    before=prefix_scale + original_score,
                    after=prefix_scale + best_score,
                    old_base=hex(original.base),
                    old_bits=list(original.bits),
                    new_base=hex(best.base),
                    new_bits=list(best.bits),
                )
                history.append(change)
                if report:
                    report(change)
            schedule[position] = best
            prob, (peak, _, _) = engine.step_and_summary(prob, left, right, best)
            prob *= 1 / peak
            prefix_scale += math.log2(peak)
            left, right = best, left
        after = prefix_scale + math.log2(endpoint_value(prob, left, right, endpoint))
        if after < before - 1e-10:
            raise AssertionError("coordinate sweep decreased the target probability")
        if report:
            report(dict(sweep=sweep + 1, before=before, after=after, accepted=accepted))
        del messages, prob
        if not accepted:
            break
    return schedule, history


def write_replay(
    engine, config, initial, schedule, endpoint, history, output, search_seconds
):
    output.parent.mkdir(parents=True, exist_ok=True)
    with output.open("x") as file:
        file.write(
            json.dumps(
                dict(
                    config=config,
                    algorithm="endpoint coordinate ascent with adjoint suffix",
                    endpoint=[
                        hex(v)
                        for v in (
                            endpoint[::-1] if config["mode"] == "linear" else endpoint
                        )
                    ],
                    changes=history,
                    optimization_seconds=search_seconds,
                    kernel_sha256=hashlib.sha256(
                        Path(__file__).with_name("kernels.cu").read_bytes()
                    ).hexdigest(),
                )
            )
            + "\n"
        )
        prob = cp.ones((1, 1), dtype=cp.float64)
        left, right = initial
        scale = 0.0
        for number, target in enumerate(schedule, 1):
            start = time.perf_counter()
            prob, (peak, total, index) = engine.step_and_summary(
                prob, left, right, target
            )
            prob *= 1 / peak
            scale += math.log2(peak)
            row, col = divmod(index, prob.shape[1])
            physical = [int(target.values()[row]), int(left.values()[col])]
            if config["mode"] == "linear":
                physical.reverse()
            cp.cuda.Stream.null.synchronize()
            record = dict(
                round=number,
                log2_max=scale,
                log2_mass=scale + math.log2(total / peak),
                output=[hex(v) for v in physical],
                window_base=hex(target.base),
                window_bits=list(target.bits),
                seconds=time.perf_counter() - start,
            )
            if number == len(schedule):
                record["log2_target"] = scale + math.log2(
                    endpoint_value(prob, target, left, endpoint)
                )
            file.write(json.dumps(record, allow_nan=False) + "\n")
            left, right = target, left
        file.write(json.dumps(dict(completed_rounds=len(schedule))) + "\n")


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("reference", type=Path)
    parser.add_argument("--device", type=int, default=0)
    parser.add_argument("--cache", choices=("host", "device"), default="device")
    parser.add_argument("--host-memory-gib", type=float, default=128)
    parser.add_argument("--candidates", type=int, default=8)
    parser.add_argument("--passes", type=int, default=3)
    parser.add_argument("--memory-gib", type=float, default=16)
    parser.add_argument("--output", type=Path, required=True)
    args = parser.parse_args()
    records = [json.loads(row) for row in args.reference.read_text().splitlines()]
    config = records[0]["config"].copy()
    rows = [r for r in records if "round" in r]
    schedule = [Window(int(r["window_base"], 0), tuple(r["window_bits"])) for r in rows]
    if (
        not rows
        or config["width"] > 20
        or not 1 <= args.candidates <= 64
        or args.passes < 1
        or args.memory_gib <= 0
        or args.host_memory_gib <= 0
    ):
        parser.error(
            "requires a nonempty width <=20 schedule, 1..64 candidates, positive passes/budget"
        )
    if args.output.exists():
        parser.error("output already exists")
    cp.cuda.Device(args.device).use()
    cp.get_default_memory_pool().set_limit(size=int(args.memory_gib * 2**30))
    engine = Engine(
        config["word_bits"], config["cipher"], config["mode"], args.memory_gib
    )
    initial = tuple(
        map(
            Window,
            (config["right"], config["left"])
            if config["mode"] == "linear"
            else (config["left"], config["right"]),
        )
    )
    endpoint = tuple(int(v, 0) for v in rows[-1]["output"])
    if config["mode"] == "linear":
        endpoint = endpoint[::-1]
    needed = sum(
        (1 << len(windows_at(initial, schedule, i)[0].bits))
        * (1 << len(windows_at(initial, schedule, i)[1].bits))
        * 8
        for i in range(len(schedule) + 1)
    )
    device_cache = needed if args.cache == "device" else 0
    if (
        device_cache * 1.3 + 6 * 8 * 2 ** (2 * config["width"])
        > args.memory_gib * 2**30
    ):
        parser.error("backward cache plus workspace exceeds conservative GPU budget")
    if args.cache == "host" and needed > args.host_memory_gib * 2**30:
        parser.error("backward cache exceeds host memory budget")
    start = time.perf_counter()
    schedule, history = optimize(
        engine,
        initial,
        schedule,
        endpoint,
        config["width"],
        args.candidates,
        args.passes,
        report=lambda record: print(json.dumps(record), flush=True),
        cache=args.cache,
    )
    elapsed = time.perf_counter() - start
    config.update(
        device=args.device,
        rounds=len(schedule),
        output=str(args.output),
        reference=str(args.reference),
        candidates=args.candidates,
        passes=args.passes,
        memory_gib=args.memory_gib,
        cache=args.cache,
        host_memory_gib=args.host_memory_gib,
        kernel="coset_lut",
        statistics="hierarchical",
    )
    write_replay(
        engine, config, initial, schedule, endpoint, history, args.output, elapsed
    )


if __name__ == "__main__":
    main()

File: src/refine.rs
//! Exact endpoint scoring: three local transitions followed by an adjoint suffix.
use crate::{
    engine::{Engine, Matrix},
    model::{Plan, Window},
    run::{self, Options},
};
use anyhow::{ensure, Result};
use serde_json::{json, Value};
use std::{path::Path, time::Instant};
struct Message {
    values: Vec<f64>,
    rows: usize,
    cols: usize,
    scale: f64,
}
fn windows(initial: &(Window, Window), schedule: &[Window], completed: usize) -> (Window, Window) {
    let left = if completed == 0 {
        initial.0.clone()
    } else {
        schedule[completed - 1].clone()
    };
    let right = if completed == 0 {
        initial.1.clone()
    } else if completed == 1 {
        initial.0.clone()
    } else {
        schedule[completed - 2].clone()
    };
    (left, right)
}
fn endpoint(prob: &Matrix, left: &Window, right: &Window, point: (u64, u64)) -> Result<f64> {
    match (left.index(point.0), right.index(point.1)) {
        (Some(i), Some(j)) => prob.value(i, j),
        _ => Ok(0.),
    }
}
fn backward(
    engine: &mut Engine,
    initial: &(Window, Window),
    schedule: &[Window],
    point: (u64, u64),
) -> Result<Vec<Message>> {
    let (left, right) = windows(initial, schedule, schedule.len());
    let i = left
        .index(point.0)
        .ok_or_else(|| anyhow::anyhow!("endpoint outside window"))?;
    let j = right
        .index(point.1)
        .ok_or_else(|| anyhow::anyhow!("endpoint outside window"))?;
    let mut prob = engine.point(left.size(), right.size(), i * right.size() + j)?;
    let mut scale = 0.;
    let mut messages = vec![Message {
        values: prob.buffer.download(prob.size())?,
        rows: prob.rows,
        cols: prob.cols,
        scale,
    }];
    for completed in (0..schedule.len()).rev() {
        let (left, right) = windows(initial, schedule, completed);
        prob = engine.adjoint(&prob, &left, &right, &schedule[completed])?;
        let stats = engine.summary(prob.ptr(), prob.size())?;
        ensure!(stats.peak > 0., "endpoint has no retained paths");
        engine.scale(&prob, 1. / stats.peak)?;
        scale += stats.peak.log2();
        messages.push(Message {
            values: prob.buffer.download(prob.size())?,
            rows: prob.rows,
            cols: prob.cols,
            scale,
        });
    }
    messages.reverse();
    Ok(messages)
}
fn score(
    engine: &mut Engine,
    prob: &Matrix,
    state: (&Window, &Window),
    tail: (&[Window], usize),
    target: &Window,
    suffix: Option<(&Matrix, f64)>,
    point: (u64, u64),
) -> Result<f64> {
    let (left, right) = state;
    let (schedule, position) = tail;
    let stop = (position + 3).min(schedule.len());
    let mut value = None;
    let (mut left, mut right) = (left.clone(), right.clone());
    let mut scale = 0.;
    for (index, next) in schedule.iter().enumerate().take(stop).skip(position) {
        let target = if index == position { target } else { next };
        let (next, stats) =
            engine.step(value.as_ref().unwrap_or(prob), &left, &right, target, None)?;
        if stats.peak <= 0. {
            return Ok(f64::NEG_INFINITY);
        }
        engine.scale(&next, 1. / stats.peak)?;
        scale += stats.peak.log2();
        value = Some(next);
        right = left;
        left = target.clone();
    }
    let mass = if let Some((suffix, suffix_scale)) = suffix {
        scale += suffix_scale;
        engine.dot(value.as_ref().unwrap(), suffix)?
    } else {
        endpoint(value.as_ref().unwrap(), &left, &right, point)?
    };
    Ok(if mass > 0. {
        scale + mass.log2()
    } else {
        f64::NEG_INFINITY
    })
}
pub fn execute(
    plan: &Plan,
    device: i32,
    memory: f64,
    host_memory: f64,
    candidates: usize,
    passes: usize,
    output: &Path,
) -> Result<()> {
    ensure!(!output.exists(), "output exists");
    ensure!(
        (1..=64).contains(&candidates) && passes > 0 && host_memory.is_finite() && host_memory > 0.,
        "invalid refinement parameters"
    );
    let initial = plan.config.initial();
    let point = plan.config.physical(plan.endpoint()?);
    let mut schedule = plan.windows.clone();
    let cache_bytes: usize = (0..=schedule.len())
        .map(|i| {
            let (l, r) = windows(&initial, &schedule, i);
            l.size() * r.size() * 8
        })
        .sum();
    ensure!(
        cache_bytes as f64 <= host_memory * 1073741824.,
        "adjoint cache exceeds host memory budget"
    );
    let workspace = 6usize * 8 * (1usize << (2 * plan.config.width));
    ensure!(
        workspace as f64 <= memory * 1073741824.,
        "conservative GPU workspace estimate exceeds budget"
    );
    let started = Instant::now();
    let mut engine = Engine::new(&plan.config, device, memory)?;
    let mut history = vec![];
    for sweep in 0..passes {
        let messages = backward(&mut engine, &initial, &schedule, point)?;
        let before = messages[0].scale + messages[0].values[0].log2();
        let mut prob = engine.point(1, 1, 0)?;
        let (mut left, mut right) = initial.clone();
        let mut prefix_scale = 0.;
        let mut accepted = 0;
        for position in 0..schedule.len() {
            let original = schedule[position].clone();
            let proposed =
                engine.candidates(&prob, &left, &right, plan.config.width, None, candidates)?;
            let mut choices = vec![original.clone()];
            for candidate in proposed {
                if !choices.contains(&candidate) {
                    choices.push(candidate);
                }
            }
            let stop = (position + 3).min(schedule.len());
            let suffix = if stop < schedule.len() {
                let m = &messages[stop];
                Some((engine.from_host(m.rows, m.cols, &m.values)?, m.scale))
            } else {
                None
            };
            let mut best = 0;
            let mut best_score = f64::NEG_INFINITY;
            let mut original_score = 0.;
            for (i, candidate) in choices.iter().enumerate() {
                let current = score(
                    &mut engine,
                    &prob,
                    (&left, &right),
                    (&schedule, position),
                    candidate,
                    suffix.as_ref().map(|(m, s)| (m, *s)),
                    point,
                )?;
                if i == 0 {
                    original_score = current;
                }
                if current > best_score + 1e-12 {
                    best = i;
                    best_score = current;
                }
            }
            let target = choices.swap_remove(best);
            if target != original {
                accepted += 1;
                let change = json!({"sweep":sweep+1,"round":position+1,"before":prefix_scale+original_score,"after":prefix_scale+best_score,"old":original,"new":target});
                println!("{change}");
                history.push(change);
            }
            schedule[position] = target.clone();
            let (next, stats) = engine.step(&prob, &left, &right, &target, None)?;
            ensure!(stats.peak > 0., "empty optimized distribution");
            engine.scale(&next, 1. / stats.peak)?;
            prefix_scale += stats.peak.log2();
            prob = next;
            right = left;
            left = target;
        }
        let after = prefix_scale + endpoint(&prob, &left, &right, point)?.log2();
        ensure!(
            after >= before - 1e-10,
            "refinement decreased endpoint probability"
        );
        let summary = json!({"sweep":sweep+1,"before":before,"after":after,"accepted":accepted});
        println!("{summary}");
        history.push(summary);
        if accepted == 0 {
            break;
        }
    }
    let seconds = started.elapsed().as_secs_f64();
    drop(engine);
    let optimized = Plan {
        config: plan.config.clone(),
        windows: schedule,
        records: vec![],
    };
    let options = Options {
        devices: vec![device],
        memory_gib: memory,
        candidates: 1,
        objective: "mass".into(),
        extend: false,
        expand: false,
    };
    run::execute(&optimized.config, Some(&optimized), &options, output)?;
    let text = std::fs::read_to_string(output)?;
    let mut records: Vec<Value> = text
        .lines()
        .map(serde_json::from_str)
        .collect::<std::result::Result<_, _>>()?;
    records[0]["optimization"] = json!({"method":"exact endpoint adjoint coordinate ascent","endpoint":plan.endpoint()?,"seconds":seconds,"host_cache_bytes":cache_bytes,"history":history});
    let data = records
        .iter()
        .map(|r| serde_json::to_string(r).unwrap())
        .collect::<Vec<_>>()
        .join("\n")
        + "\n";
    std::fs::write(output, data)?;
    Ok(())
}
